src/agent/vsoworker.ts

```diff
@@ -133,26 +149,37 @@ export function run(msg, consoleOutput: boolean,
                 wk.error('Error: ' + err.message);
             }
 
-            ctx.finishJob(result, (err: any) => {
-                trace.callback('ctx.finishJob');
- 
-                wk.status('Job Finished: ' + job.jobName);
-                if (err) {
-                    wk.error('Error: ' + err.message);
-                }
-
-                finished();
-            });
+            // if finishingJob is true here, then the lock renewer got a 404, which means the server abandoned the job
+            // it's already calling serviceChannel.drain() and it's going to call finished()
+            if (!finishingJob) {
+                finishingJob = true;
+                ctx.finishJob(result).fin(() => {
+                    // trace and status no matter what. if finishJob failed, the fail handler below will be called
+                    trace.callback('ctx.finishJob');
+                    wk.status('Job Finished: ' + job.jobName);
+                }).fail((err: any) => {
+                    if (err) {
+                        wk.error('Error: ' + err.message);
+                    }
+                }).fin(() => {
+                    deferred.resolve(null);
+                });
+            }
         });
     }
+    else {
```

Comment:
Previously, if the message was not a "job", the finished() callback would never be invoked.  I suspect that was not the desired behavior...